### ISSUE TITLE
Brainstorming: Improving address mapping and validation

### ISSUE THREAD

**Issue by tordans** [openstreetmap/iD#6177]
Inspired by a presentation on a German OSM(…) conference ([YouTube, Audio in German](https://www.youtube.com/watch?v=AIBi2aD-4n0)), I looked at https://addresses.lorenz.lu/oad_dach/countries which is a tool about address validation in OSM.

There are a few use cases that could be handled by iD to improve address data in OSM.

This ticket is meant more inspirational, than a specific feature request. Especially since some of the goals can be solved in different ways. I suggest we collect ideas here and create separate tickets once we haven a clearer idea what is worth solving ATM.

**UseCase: Copy paste an address (as node or when copy-pasting a house).**
I copy paste houses all the time since its less clicks to copy+move+rotate than to draw them from scratch. This has a higher risk of copying the address as well.
- iD could validate „is there another house in the viewport with the same address, if yes, suggest checking it“. Actions would be „change address = edit object“ or „ignore“

**UseCase: Copy paste a house which has an address to another road nearby, then forget to change the road.**
  - iD could validate „if the house that I edited has a street name, check if the street nearby has the same name. If not, show a warning.“ Actions would be „copy street name from street to the house“, „edit address = edit object“, „ignore“

**UseCase: I create a new house next to a road but do not add an address.**
- iD could provide a feature, saying „copy street name from the road nearby“ (just the street name).
- And in case, there is a house nearby with address: „copy address (without number) from the house next to this one“ (street name, zip, city name).
- Those feature could be part of the validator, or a separate notice / action / button on the address-preset-section.
  - Or maybe both, a separate feature but also a validator.

- Another take on this, where to add those address validations via an API to https://github.com/britiger/osm_address_db/issues/18. Or to add the validation rules to KeepRight instead. This would not provide live feedback, but at least show the issues inside the editor.

**UseCase: I copy a house with a name attribute.**
- iD could validate the name the same way as the house number (see above). This is interesting for all attributes that should usually not be copy-pasted. So also URLs, Owner. What else?
- Maybe the preset-JSON is a place where we mark those features that should be validated on copy-paste actions since they should be unique? (I would not remove them automatically on copy-paste.)

**UseCase: Complete Address street … after adding the number on the go**
- I use GoMap to add street numbers on the go ([Example](https://www.openstreetmap.org/node/6396864204)). However, re-adding the rest of the street data is way to tedious in GoMap or with any iOS App (see also https://github.com/bryceco/GoMap/issues/19). So Ideally, I could use GoMap to place the street-number and iD would help me with some live validation to add other data. With a copy-button as described above.

**Comment by Adamant36** [comment]
> iD could validate „if the house that I edited has a street name, check if the street nearby has the same name. If not, show a warning.“ Actions would be „copy street name from street to the house“, „edit address = edit object“, „ignore“

What if the house is on a corner and its not clear what street its facing? It might encourage people to add the wrong street name. Sometimes the drop street down does that.

**Comment by bhousel** [comment]
> I copy paste houses all the time since its less clicks to copy+move+rotate than to draw them from scratch. This has a higher risk of copying the address as well.

Yes, iD originally omitted the name, address, contact details when copy-pasting, but in #4067 people asked for this, so I changed my mind and now we let people copy them.

I'd be ok with adding an address validator to catch common mistakes (like duplicates).

The elephant in the room - OSM is kind of not great at addresses in the first place.  There are multiple competing methods to map them, which makes it very hard for software to work with them.  Also there is this [trove of open address data](https://openaddresses.io/) that the OSM community refuses to use.  I think about this sometimes, and wonder if maybe we should just say "addresses don't belong in OSM"  (like some kinds of historical data, or parcel data, or some boundaries).  Or at least acknowledge that there are much bigger barriers to making OSM a great source of address data than whatever iD does.

**Comment by slhh** [comment]
When pasting an object, there is the phase between pressing Ctrl-V and placing the new object. In this phase a context dependent top bar (and also shortcuts) can give the user options to specify the extend of copying.
E.g. 
1. geometry+preset
2. geometry+partially tags (without addresses etc.)
3. geometry+full tags

In addition the sidebar can show the tags to be copied and allow to enable/disable copying specific tags.

**Comment by manfredbrandl** [comment]
Addresses do belong to OSM! It would help a lot, if iD will allow to show addresses or address-validation-Info from different sources. Maybe we need another Community Index for this!

**Comment by matkoniecz** [comment]
Addresses certainly belong in OSM, at least in Poland mapping is going quite well. Both manual collection and imports where official data was published and is worth importing.

> this trove of open address data that the OSM community refuses to use

Can you link documentation page of an import that was rejected by OSM community? I am curious what went wrong. Or is it case of "nobody so far had time to handle it".

**Comment by ArenaL5** [comment]
I wish to contribute with a different use case that is related to addresses:

![Address instead of entrance](https://user-images.githubusercontent.com/37388405/59923758-b2709680-9423-11e9-923a-a5b6565ae5f6.png)

Lots of people (in northwestern Spain, at least) omit the key `entrance=yes` or `entrance=home` when adding a house number at the edge of a building. Could the validator suggest to add this key? There's little reason to add an address with no other tags like in the screenshot.
